### PR TITLE
Add logic for permissions roles and assignment to the domain and persistence layer

### PR DESCRIPTION
**Pull request by kstam**

### CODE REVIEW

File: src/domain/Uuid.test.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { Uuid } from './Uuid';
```
Comment (Tankenstein) [inline src/domain/Uuid.test.ts]: Why not just export the `v4` function from the `UUID` domain? Actual reference equality is quite nice

Comment (Tankenstein) [reply src/domain/Uuid.test.ts]: Ah you want to extend it

File: src/domain/model/user/User.ts [hunk 2]
```diff
@@ -47,4 +58,27 @@ export class User {
...
+    return this.roleAssignments.activeAssignments().map((role) => role.name);
+  }
+
+  get permissions(): string[] {
+    const permissions: Set<string> = new Set();
+    this.roleAssignments.activeAssignments().forEach((role) => {
+      role
+        .permissions()
+        .map((permission) => permission.name)
+        .forEach((permissionName) => permissions.add(permissionName));
```
Comment (StevePole) [inline src/domain/model/user/User.ts]: This shouldn't need the parameter brackets, surprised the formatter isn't removing them?

Comment (kstam) [reply src/domain/model/user/User.ts]: Which brackets do you mean? Around `permissionName`? They are always added by the formatter even for single parameter lambdas

File: src/domain/model/authentication/AssignmentActions.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
+import { AssignmentAction, AssignmentId } from './AssignmentAction';
+import { AssignmentActionType } from './AssignmentActionType';
+import { UserId } from '../user/UserId';
+
+export class AssignmentActions<T, R> {
```
Comment (StevePole) [inline src/domain/model/authentication/AssignmentActions.ts]: My main question with this is whether we need the generics, or whether we could achieve the same thing with a base class?   It's very hard to follow to follow at the moment.

I would even consider a comment above the class definition.
```
// R represents the Resource, T represents what it's assigned To
export class AssignmentActions<T, R> {
```
Otherwise I need to read quite far into the class to work out what these generics represent.

File: src/domain/model/authentication/AssignmentActions.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+  addAssignment(resource: R, assignTo: T, actor: UserId) {
+    this.newAssignmentAction(
+      new AssignmentAction(
+        new AssignmentId(),
+        assignTo,
+        resource,
+        AssignmentActionType.ADD,
+        actor,
+        this.assignmentActions.length + 1
```
Comment (StevePole) [inline src/domain/model/authentication/AssignmentActions.ts]: Why do we need this?

File: src/domain/model/authentication/AssignmentActions.ts (hunk 3 above)
Comment (StevePole) [reply src/domain/model/authentication/AssignmentActions.ts]: I realise there's a convention to use single letters, but I don't believe that's a constraint.  Would you consider naming them
`export class AssignmentActions<To, Resource> {`
?

I think that would be more readable rather than less.

File: src/domain/model/authentication/Role.ts [hunk 5]
```diff
@@ -1,13 +1,45 @@
...
 export class Role {
+  readonly permissionAssignments: AssignmentActions<Role, Permission>;
+
   constructor(
-    readonly id: RoleId,
     readonly name: string,
-    readonly creationTime: Date
-  ) {}
+    readonly creationTime: Date = new Date(),
+    permissionAssignments: PermissionAssignmentAction[] = []
```
Comment (StevePole) [inline src/domain/model/authentication/Role.ts]: Would it make sense to rearrange these two params?  i.e. Might you have a scenario where you create this with assignments, or would that only happen when reading from the DB?

File: src/infrastructure/persistence/PsqlPermissionRepository.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import { PermissionRepository } from '../../domain/model/authentication/PermissionRepository';
+import { Permission } from '../../domain/model/authentication/Permission';
+import knex from 'knex';
+
+export class PsqlPermissionRepository implements PermissionRepository {
+  constructor(private db: knex) {}
+
+  async findByName(name: string): Promise<Permission | null> {
+    const permissionRow: any = await this.db('permission')
+      .select(['name', 'creation_time'])
```
Comment (StevePole) [inline src/infrastructure/persistence/PsqlPermissionRepository.ts]: as creationTime?

File: src/infrastructure/persistence/PsqlRoleRepository.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+import { AssignmentActionType } from '../../domain/model/authentication/AssignmentActionType';
+import { UserId } from '../../domain/model/user/UserId';
+
+export class PsqlRoleRepository implements RoleRepository {
+  constructor(private db: knex) {}
+
+  async findByName(name: string): Promise<Role | null> {
+    const roleRows: any[] = await this.db('role')
+      .select([
+        'role.name as role_name',
```
Comment (StevePole) [inline src/infrastructure/persistence/PsqlRoleRepository.ts]: Could do `as roleName` here as well, just to stick to JS conventions, and follow the patterns we're using elsewhere.

File: src/infrastructure/persistence/PsqlRoleRepository.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+      actor: permissionAssignmentAction.actor.value,
+      creation_time: permissionAssignmentAction.creationTime,
+      order: permissionAssignmentAction.order,
+      action_type: permissionAssignmentAction.actionType.toString(),
+    });
+  }
+}
+
+export function createRoleWithAssignedPermissions(roleRows: any[]): Role {
+  const role = new Role(roleRows[0].role_name, roleRows[0].role_creation_time);
```
Comment (StevePole) [inline src/infrastructure/persistence/PsqlRoleRepository.ts]: Why does this only use the first entry in the array?  I'm guessing that, because of stuff happening elsewhere in this file, then all of these rows have the same role?  Is it reasonable to assume that though?  Might be better to pass it in?

File: src/infrastructure/persistence/PsqlUserRepository.ts [hunk 9]
```diff
@@ -69,19 +83,111 @@ export class PsqlUserRepository implements UserRepository {
...
-    userRow.modificationTime
-  );
+  private async getRoleAssignments(
+    user: User
+  ): Promise<RoleAssignmentAction[]> {
+    const roleAssignmentRows: any[] = await this.db(
+      'role_to_user_assignment as rua'
+    )
+      .select([
+        'rua.id as rua_id',
```
Comment (StevePole) [inline src/infrastructure/persistence/PsqlUserRepository.ts]: Same comment regarding column aliasing: `as ruaId`

File: src/infrastructure/persistence/PsqlUserRepository.ts [hunk 10]
```diff
@@ -69,19 +83,111 @@ export class PsqlUserRepository implements UserRepository {
...
+      .where('rua.user_id', '=', user.id.value);
+
+    const groupedByAssignmentId = this.groupByAssignmentId(roleAssignmentRows);
+    return this.mapToRoleAssignmentActions(groupedByAssignmentId, user);
+  }
+
+  private groupByAssignmentId(roleAssignmentRows: any[]) {
+    const groupedByAssignmentId = new Map<string, any[]>();
+    roleAssignmentRows
+      .filter((row) => row.role_name && row.role_creation_time)
```
Comment (StevePole) [inline src/infrastructure/persistence/PsqlUserRepository.ts]: Don't need parenthesis again `(row) => `

File: src/infrastructure/persistence/PsqlUserRepository.ts [hunk 11]
```diff
@@ -69,19 +83,111 @@ export class PsqlUserRepository implements UserRepository {
...
+  }
+
+  private mapToRoleAssignmentActions(
+    groupedByAssignmentId: Map<string, any[]>,
+    user: User
+  ) {
+    const roleAssignmentActions: RoleAssignmentAction[] = [];
+    groupedByAssignmentId.forEach((roleRows) => {
+      const role = createRoleWithAssignedPermissions(roleRows);
+      roleAssignmentActions.push(
```
Comment (StevePole) [inline src/infrastructure/persistence/PsqlUserRepository.ts]: There's a lot going on in this statement.  I would be tempted to break it out to make it a little more readable.  Perhaps this just to remove one level of nesting:
```
const roleAssignmentAction = new RoleAssignmentAction(
  ...
);

roleAssignmentActions.push(roleAssignmentAction);
```

Comment (StevePole) [reply src/infrastructure/persistence/PsqlUserRepository.ts]: Or maybe make the creation of the single roleAssignmentAction a separate function?

File: src/domain/model/authentication/AssignmentActions.ts (hunk 3 above)
Comment (kstam) [reply src/domain/model/authentication/AssignmentActions.ts]: Yeah, I usually try to stick to the Generics naming conventions but I agree that in this case it might make more sense to follow class name conventions.

File: src/domain/model/authentication/AssignmentActions.ts (hunk 4 above)
Comment (kstam) [reply src/domain/model/authentication/AssignmentActions.ts]: This is to guarantee order when actions are added sequentially (i.e an ADD followed by a REMOVE). If you solely rely on the creationTime there is a high chance they will all have the same millisecond and you won't know which one goes first. Using the order makes sure that order is always guaranteed. (In java this would be an AtomicInt, but node has a single thread so we're good)

File: src/domain/model/authentication/Role.ts (hunk 5 above)
Comment (kstam) [reply src/domain/model/authentication/Role.ts]: Yep, this makes sense. I'll update

File: src/infrastructure/persistence/PsqlPermissionRepository.ts (hunk 6 above)
Comment (kstam) [reply src/infrastructure/persistence/PsqlPermissionRepository.ts]: Hmm.. Might have to update ALL repos if we do this 🤔

File: src/infrastructure/persistence/PsqlRoleRepository.ts (hunk 8 above)
Comment (kstam) [reply src/infrastructure/persistence/PsqlRoleRepository.ts]: Yeah, as all rows come back from DB with left join, if a row has multiple assigned permissions it will be the same in every row. I'll see how I can make a bit more readable

File: src/infrastructure/persistence/PsqlUserRepository.ts (hunk 10 above)
Comment (kstam) [reply src/infrastructure/persistence/PsqlUserRepository.ts]: We don't need this but this is how prettier is defined. try to remove it and you'll see it'll add it back. :)

File: src/infrastructure/persistence/PsqlUserRepository.ts (hunk 11 above)
Comment (kstam) [reply src/infrastructure/persistence/PsqlUserRepository.ts]: Yeah, good point. I'll move to a separate function!

File: src/infrastructure/persistence/PsqlRoleRepository.ts (hunk 8 above)
Comment (kstam) [reply src/infrastructure/persistence/PsqlRoleRepository.ts]: I think it's actually better to not change this.

The purpose of the function is to get a list of rows for the same role and create a row with all its permissions. It's better that it creates the Role

File: src/domain/model/authentication/Role.ts (hunk 5 above)
Comment (kstam) [reply src/domain/model/authentication/Role.ts]: Actually on a second thought I wouldn't change this. To assign permissions to the role you need to first create the role. So I would just restrict completely from the constructor.

File: src/infrastructure/persistence/PsqlUserRepository.ts [hunk 12]
```diff
@@ -27,13 +32,13 @@ export class PsqlUserRepository implements UserRepository {
...
-        values (:id, :email, :creation_time, :modification_time)
+        values (:id, :email, :creationTime, :modificationTime)
         on conflict do nothing`,
       {
         id: user.id.value,
         email: user.email.value,
-        creation_time: user.creationTime,
-        modification_time: user.modificationTime,
+        creationTime: user.creationTime,
+        modificationTime: user.modificationTime,
```
Comment (StevePole) [inline src/infrastructure/persistence/PsqlUserRepository.ts]: Don't think you want to change this one?

Comment (StevePole) [reply src/infrastructure/persistence/PsqlUserRepository.ts]: `modification_time: user.modificationTime,`

Comment (StevePole) [reply src/infrastructure/persistence/PsqlUserRepository.ts]: I guess it doesn't matter as you also changed in the statement

Comment (kstam) [reply src/infrastructure/persistence/PsqlUserRepository.ts]: It was actually intentional.. Just not sure where to draw the line 😇

File: src/domain/model/user/User.ts (hunk 2 above)
Comment (StevePole) [reply src/domain/model/user/User.ts]: Ok, weird, but fine.

File: src/infrastructure/persistence/PsqlUserRepository.ts (hunk 12 above)
Comment (StevePole) [reply src/infrastructure/persistence/PsqlUserRepository.ts]: In other repositories (and actually later on in this one), we use underscores on the left, and camelCase on the right.  But it's not a big deal.

Comment (StevePole) [reply src/infrastructure/persistence/PsqlUserRepository.ts]: Actually, not later on this one, it's not a raw statement.
